### PR TITLE
Repair type errors

### PR DESCRIPTION
**Pull request by NeilGirdhar**
This is a minor change to fix various type annotation errors

## Checklist
- [x] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other
      checks if that's the case).

### CODE REVIEW

File: flax/optim/dynamic_scale.py [hunk 1]
```diff
@@ -86,7 +86,7 @@ def loss_fn(p):
   def value_and_grad(self, fun: Callable[..., Any],
                      argnums: Union[int, Sequence[int]] = 0,
                      has_aux: bool = False,
-                     axis_name: str = None,
+                     axis_name: Optional[str] = None,
```
Comment (NeilGirdhar) [inline flax/optim/dynamic_scale.py:89]: This was changed in the typing PEP, and type-checkers now ask that these be explicitly marked optional

**Comment by NeilGirdhar** [comment]
It would also be really nice to get rid of:
```python
    def __init__(*args, **kwargs):
      # this stub makes sure pytype accepts constructor arguments.
      pass
    def __call__(self, *args, **kwargs) -> Any:
      # this stub allows pytype to accept Modules as Callables.
      pass
```
The `__call__` defintion causes MyPy to rightly complain for _every_ user-defined module:
```
cmm/haiku_module/linear.py:41: error: Signature of "__call__" incompatible with supertype "Module"  [override]
cmm/haiku_module/linear.py:41: note:      Superclass:
cmm/haiku_module/linear.py:41: note:          def __call__(self, *args: Any, **kwargs: Any) -> Any
cmm/haiku_module/linear.py:41: note:      Subclass:
cmm/haiku_module/linear.py:41: note:          def __call__(self, inputs: ndarray[Any, dtype[floating[Any]]]) -> ndarray[Any, dtype[floating[Any]]]
```
There must be an easier way to accomplish whatever it is that's trying to be accomplished with pytype?

**Comment by NeilGirdhar** [comment]
(Not sure I can't reply to the review in line.)

For the comment about `Set[str]`, the cleanest would be in Python 3.9 to do `self.reservations = set[str]()`.  I'll move the annotation to a class level annotation if that seems cleaner to you.

As for the `get_variable`, I guess you could assert that the return value is a subclass of the default, but I don't think that's what you want?  The default could be None, and then I assume you don't expect `None` as the only possible return value?

**Comment by NeilGirdhar** [comment]
@jheek Hey, I'm just about to rebase this again.   It's been a few weeks, any chance we could get this in soon so I don't have to keep rebasing it?

File: flax/optim/dynamic_scale.py (hunk 1 above)
Comment (jheek) [reply flax/optim/dynamic_scale.py]: Makes sense

**Review by jheek — APPROVED** [review]
Thanks!

**Comment by NeilGirdhar** [comment]
@marcvanzee Any idea how I fix this?  Since I'm not a Googler, I can't see the internal failure.

**Comment by marcvanzee** [comment]
@jheek there seem to be internal tests failing related to this PR. Could you take a look?

**Comment by NeilGirdhar** [comment]
I've rebased this off the latest pushes.

**Comment by jheek** [comment]
@NeilGirdhar we had some internal issues with this PR that I fixed internally together with the merge conflicts.

**Comment by NeilGirdhar** [comment]
@jheek Great, thank you!

**Pull request merged**